src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -19,80 +23,212 @@
+            )
+        elif passphrase != passphrase2:
+            for f in ['passphrase', 'passphrase2']:
+                verrors.add(
+                    f'{schema_name}.{f}',
+                    'Pass phrase confirmation does not match'
+                )
+    elif passphrase or passphrase2:
+        verrors.add(
+            f'{schema_name}.passphrase',
+            'No private key specified for pass phrase'
+        )
+
+    if (
+        (certificate and private_key) and
+        all(k not in verrors for k in (f'{schema_name}.certificate', f'{schema_name}.privatekey'))
+    ):
+        public_key_obj = crypto.load_certificate(crypto.FILETYPE_PEM, certificate)
+        private_key_obj = await middleware.call('certificate.load_private_key', private_key, passphrase)
 
-        # getting fingerprint of certificate
         try:
-            certificate = crypto.load_certificate(
-                crypto.FILETYPE_PEM,
-                cert_certificate
+            context = SSL.Context(SSL.TLSv1_2_METHOD)
+            context.use_certificate(public_key_obj)
+            context.use_privatekey(private_key_obj)
+            context.check_privatekey()
+        except SSL.Error as e:
+            verrors.add(
+                f'{schema_name}.privatekey',
+                f'Private key does not match certificate: {e}'
             )
-        except Exception:
-            return None
-        else:
-            return certificate.digest('sha1').decode()
+
+    key_length = data.get('key_length')
+    if key_length:
+        if key_length not in [1024, 2048, 4096]:
+            verrors.add(
+                f'{schema_name}.key_length',
+                'Key length must be a valid value ( 1024, 2048, 4096 )'
+            )
+
+    signedby = data.get('signedby')
+    if signedby:
+        valid_signing_cas = await middleware.call(
+            'certificateauthority.query',
+            [
+                ('certificate', '!=', None),
+                ('privatekey', '!=', None),
+                ('certificate', '!=', ''),
+                ('privatekey', '!=', ''),
+            ],
+        )
+
+        if signedby not in [d['id'] for d in valid_signing_cas]:
+            verrors.add(
+                f'{schema_name}.signedby',
+                'Please provide a valid signing authority'
+            )
+
+
+class CertificateService(CRUDService):
+
+    class Config:
+        datastore = 'system.certificate'
+        datastore_extend = 'certificate.cert_extend'
+        datastore_prefix = 'cert_'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.map_functions = {
+            'CERTIFICATE_CREATE_INTERNAL': self.__create_internal,
+            'CERTIFICATE_CREATE_IMPORTED': self.__create_imported_certificate,
+            'CERTIFICATE_CREATE': self.__create_certificate,
+            'CERTIFICATE_CREATE_CSR': self.__create_csr
+        }
+
+    @filterable
```

Comment:
It seems this whole query method could be removed now, right? Since its implemented in base CRUDService

Reply:
True, it is not needed anymore, I'll revise this